### ISSUE TITLE
ENH: add Chandrasekhar recursions estimator for SARMA models

### ISSUE THREAD

**Issue by ChadFulton** [statsmodels/statsmodels#6812]
As noted on the [mailing list](https://groups.google.com/forum/#!topic/pystatsmodels/1Oq1jYbrHgY), recent versions of Gretl estimate SARMA models very quickly, apparently using the methods described in Melard (1984) "Algorithm AS 197: A Fast Algorithm for the Exact Likelihood of Autoregressive-Moving Average Models".

This algorithm is essentially an application of using the Chandrasekhar recursions rather than the Kalman filtering recursions, as we added in #6411, but it is specialized to the ARMA case and apparently is very fast.

I think it should be added along as an estimator to the new `ARIMA` model in the same way as was done for `innovations_mle`. So it would live at e.g. `statsmodels/tsa/arima/estimators/chandrasekhar`.

**Comment by w31ha0** [comment]
It seems like the previous PR #6411 has the kalman algo already written in a different folder (_statespace_/_filters_) and in a pxd format instead of standard python.

Does this mean that for this issue, one is expected to write the Chandrasekhar recursion from scratch in the new place and in standard python? (statsmodels/tsa/arima/estimators/chandrasekhar)

**Comment by ChadFulton** [comment]
Yes, the PR you mentioned added General Chandrasekhar recursions for state space models, but this issue is about implementing the specific algorithm in the reference paper. Although it also uses Chandrasekhar recursions, that paper has considered how to optimize them for the case of ARIMA models.

(Alternatively, maybe it’s possible to optimize the Chandrasekhar recursions for general state space models, but I think that would be harder than just implementing the algorithm in the paper for ARIMA models specifically).

The pyx and pxd files are Cython files, which compile to C for improved performance. It’s always best to prototype in Python, though, so if you had interest doing that, that would be great too. Eventually, we’ll probably have to either use Cython (or maybe Numba), for performance reasons.

Finally, the paper makes available some Fortran code that might be of use somehow (I think the Greek implementation uses that directly) but (a) I’m not sure if the license is compatible, and (b) we’d still have to translate it to Cython, since I don’t think we want to handle Fortran sources in our build process.

**Comment by w31ha0** [comment]
I'm interested to tackle this but I'm struggling to understand how equations 5 - 9 in the paper "Algorithm AS 197: A Fast Algorithm for the Exact Likelihood of Autoregressive-Moving Average Models" are derived. It seems like these details could be found in "Some new algorithms for recursive estimation in constant linear systems" but I can't really find this paper anywhere too.

**Comment by ChadFulton** [comment]
If you're interested in more details on how the Chandrasekhar recursions work for state space models in general, one option is to look at "Using the “Chandrasekhar Recursions” for Likelihood
Evaluation of DSGE Models" (Herbst, 2015), which can be found at https://link.springer.com/content/pdf/10.1007/s10614-014-9430-2.pdf
